src/ray/core_worker/core_worker.cc

```diff
@@ -228,6 +228,16 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
   profiler_ = std::make_shared<worker::Profiler>(
       worker_context_, options_.node_ip_address, io_service_, gcs_client_);
 
+  // Initialize the task state event buffer.
+  if (RayConfig::instance().task_events_report_interval_ms() > 0) {
+    auto task_event_gcs_client = std::make_unique<gcs::GcsClient>(options_.gcs_options);
+    task_event_buffer_ =
```

Comment:
(and we can make API no-op when the report interval == 0?)

Reply:
Hmm, I wonder if we should not be passing raw pointers around? 

I guess we will need to check if the feature is on (or if the TaskEvenBuffer is setup properly) either at the callsite (which I am doing here) or in the TaskEventBuffer respective functions. 

So in terms readability, we will do a `if-not-on` clause somewhere regardless it being a pointer or concrete instance. I thought it would make more sense to explicitly show the feature being off by having a nullpointer handle.

Reply:
Re; unique vs shared pointer I think since the pointer's lifetime is explicitly managed by the core worker, it is okay to pass the raw pointer. It's actually easier to understand the lifecycle of the pointer (the owner of the unique pointer can easily guarantee all child classes holding a raw pointer is safe) cc @iycheng (he usually prefers this policy). 

> So in terms readability, we will do a if-not-on clause somewhere regardless it being a pointer or concrete instance. I thought it would make more sense to explicitly show the feature being off by having a nullpointer handle.

I will leave it up to you for the fast iteration... (so feel free to go whatever you want). In my perspective readability is not so important here though. For example, I wouldn't care if this feature is enabled or not when I read/write code. Since that should be handled at the higher layer (task_events_report_interval_ms). The con of the current style is that it is more error-prone and add cognitive load to API users. I think no-op is similar to having multiple implementation with the default impl NullOperation (which is also pretty common).

Reply:
> it is okay to pass the raw pointer. It's actually easier to understand the lifecycle of the pointer (the owner of the unique pointer can easily guarantee all child classes holding a raw pointer is safe) 

Sounds good to me - if we don't follow a no-raw-pointer-policy.

Reply:
Kept the nullptr check at callsite, rather than having no-op in `TaskEventBuffer` funcs so that prep of arguments will be avoided.